### PR TITLE
[REVIEW] Port unary cast to libcudf++

### PR DESCRIPTION
**Pull request by trxcllnt**
Port unaryops `cast` for issue #2950, related to @codereport's PR https://github.com/rapidsai/cudf/pull/3214.

### CODE REVIEW

**Comment by trxcllnt** [comment]
rerun tests

File: cpp/include/cudf/unary.hpp [hunk 1]
```diff
@@ -16,8 +16,9 @@
 
 #pragma once 
 
-#include <cudf/cudf.h>
+#include <memory>
 #include <cudf/types.hpp>
+#include <rmm/mr/default_memory_resource.hpp>
```
Comment (jrhemstad) [inline cpp/include/cudf/unary.hpp]: The forward declarations in `types.hpp` should be sufficient. 
```suggestion
```

File: cpp/src/unary/cast_ops.cu [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+std::unique_ptr<column> cast(column_view const& input,
+                             data_type type,
+                             rmm::mr::device_memory_resource* mr) {
+  CUDF_EXPECTS(is_fixed_width(type), "Unary cast type must be fixed-width.");
+
+  auto size = input.size();
+  auto null_mask = copy_bitmask(input, 0, mr);
+  auto output = std::make_unique<column>(
+      type, size, rmm::device_buffer{size * cudf::size_of(type), 0, mr},
+      null_mask, input.null_count(), std::vector<std::unique_ptr<column>>{});
```
Comment (jrhemstad) [inline cpp/src/unary/cast_ops.cu]: Careful, this was deep copying the device buffer.
```suggestion
      std::move(null_mask), input.null_count(), std::vector<std::unique_ptr<column>>{});
```

Comment (jrhemstad) [inline cpp/src/unary/cast_ops.cu]: The default arg should be sufficient. 
```suggestion
      null_mask, input.null_count());
```

File: cpp/src/unary/cast_ops.cu [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  void operator()(cudaStream_t stream) {
+    experimental::type_dispatcher(
+        output.type(), dispatch_unary_cast_to<T>{input, output}, stream);
+  }
+
+  template <typename T,
+            typename std::enable_if_t<!cudf::is_timestamp<T>() &&
+                                      !cudf::is_numeric<T>()>* = nullptr>
+  void operator()(cudaStream_t stream) {
+    CUDF_FAIL("Column type must be numeric");
```
Comment (jrhemstad) [inline cpp/src/unary/cast_ops.cu]:
```suggestion
    CUDF_FAIL("Column type must be numeric or timestamp");
```

File: cpp/src/unary/cast_ops.cu (hunk 2 above)
Comment (trxcllnt) [reply cpp/src/unary/cast_ops.cu]: Don't we want to deep-copy the device buffer? What happens if the input column is still valid and in use elsewhere?

**Comment by trxcllnt** [comment]
@jrhemstad I cleaned up the implementations and changed the signatures to return `unique_ptr<column>` now that https://github.com/rapidsai/cudf/pull/3232 is merged, I was thinking of updating this PR to call `to_timestamps` and `from_timestamps` when casting TS to/from strings. Does that sound like the right approach?

**Comment by jrhemstad** [comment]
> I was thinking of updating this PR to call `to_timestamps` and `from_timestamps` when casting TS to/from strings. Does that sound like the right approach?

If that's the expected behavior of casting to/from strings, then sure.

**Comment by trxcllnt** [comment]
It would be convenient to have a single `cudf::cast()` method to convert columns to another type, but after looking through the methods in `cudf/strings/convert`, I've decided against it.

Each string method takes different parameters based on the to/from type, and it isn't possible to define `cast()` overloads for each convert func parameter list. Better to guide people to those functions explicitly when they want to go to/from strings.

**Comment by harrism** [comment]
What's the status of this PR?

File: cpp/include/cudf/unary.hpp [hunk 4]
```diff
@@ -42,5 +42,17 @@ std::unique_ptr<cudf::column> is_null(cudf::column_view const& input);
  */
 std::unique_ptr<cudf::column> is_valid(cudf::column_view const& input);
 
+/**
+ * @brief  Casts data from dtype specified in input to dtype specified in output
+ *
```
Comment (jrhemstad) [inline cpp/include/cudf/unary.hpp]: Might want to add some documentation about what types are supported.

**Comment by trxcllnt** [comment]
@harrism to answer your question, now that I've updated the `cast` docstring I think this is ready to merge.

**Pull request merged**